### PR TITLE
Replace websocketsharp with Websocket.Client

### PR DESCRIPTION
**Pull request by Craig-J**
## Summary

_Summary of the purpose of this pull request._

Replaces websocketsharp with [Websocket.Client](https://github.com/Marfusios/websocket-client).

Websocketsharp has been a source of several bugs including a performance-sensitive stack overflow on receives causing debugging issues and issues getting into an inconsistent state when the connection is closed by the remote.

Closes [TF-744](https://ultrahaptics.atlassian.net/browse/TF-744)
Closes [TF-1095](https://ultrahaptics.atlassian.net/browse/TF-1095)
Closes [TF-1137](https://ultrahaptics.atlassian.net/browse/TF-1137)

### Tests Added

_Summary of the tests added as part of this change, or reason why no tests are required._

No automated tests have been added in this PR. Whether they should be required or not is open to discussion. I initially explored adding tests to check basic functionality before quickly realising I would just be testing the `WebSocket` library itself and not our usage of it. If we want to add any tests they should be integration tests with a real or mock Diagnostic API. Question is whether these would be valuable long term or whether some manual testing during this PR is sufficient.

## Contributor Tasks

_These tasks are for the pull request creator to tick off._

- [ ] PO review (optional depending on work type)
- [ ] XDR review (optional depending on work type)
- [ ] QA review (or another developer if no QA is available)
- [x] Ensure documentation requirements are met e.g., public API is commented
- [x] Relevant changelogs have been updated with user-visible changes
    - [x] [TouchFree Windows](/ultraleap/touchfree/blob/-/CHANGELOG-windows.md)
    - [x] [TouchFree BrightSign](/ultraleap/touchfree/blob/-/CHANGELOG-brightsign.md)
    - [ ] [TouchFree Web Tooling](/ultraleap/touchfree/blob/-/TF_Tooling_Web/CHANGELOG.md)
- [x] Consider any licensing/other legal implications e.g., notices required

If there is an associated JIRA issue:
- [x] Include a link to the JIRA issue in the summary above
- [x] Make sure the fix version on the issue is set correctly

## Reviewer Tasks

_Add any instructions or tasks for the reviewer such as specific test considerations before this can be merged._

- [x] Developer testing
- [x] Code reviewed
- [ ] Non-code assets reviewed
- [ ] Documentation reviewed - includes checking documentation requirements are met and not missing e.g., public API is commented

[TF-744]: https://ultrahaptics.atlassian.net/browse/TF-744?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TF-1095]: https://ultrahaptics.atlassian.net/browse/TF-1095?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TF-1137]: https://ultrahaptics.atlassian.net/browse/TF-1137?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Craig-J** [comment]
- [x] Can get stuck disconnected when no connection was make successfully initially e.g. Service not connected at all.
- [x] Response accumulation behaviour in `TrackingApiChangeQueueHandler` blocks newly queued requests that would cause a reconnect attempt from recovering from the issue sooner. Existing timeout for the `trackingApiResponse` makes this state recoverable but the timeout is currently a long time.
- [x] Can get stuck disconnected until `ErrorReconnectionTimeout` passes (default 60s) when a reconnection fails due to an error connecting. Any new requests relating to diagnostic API should cause a reconnection attempt. This is blocked partially by the issue - unclear currently if that's the only issue preventing a quicker recovery.

**Comment by DavidFee-UL** [comment]
Craig - I checked out the code and certainly the stack overflow is fixed. Few observations:

- Did you use 'git mv' when moving the *TrackingDiagnostic* files? Not sure how github works here but it looks like the file's been deleted and then created. It means the change history gets lost for the file.

- Looking at your comments on testing, agree we wouldn't want to be testing the websocket library, and your [comments](https://ultrahaptics.atlassian.net/browse/TF-1095?focusedCommentId=88629) show it's very hard to get it to fail. Beyond debugging the library I agree it's very difficult to test this.  Frustrating...

- A number of TODO comments in the code - checked with the team on best practice, hence my comments.

Probably needs someone with more TouchFree C# experience to review the specifics of the code changes TBH.

**Comment by DavidFee-UL** [comment]
Looking at your comments on testing, agree we wouldn't want to be testing the websocket library, and your [comments](https://ultrahaptics.atlassian.net/browse/TF-1095?focusedCommentId=88629) show it's very hard to get it to fail. Beyond debugging the library I agree it's very difficult to test this.  Frustrating...

File: TF_Service_dotNet/TouchFree/Connections/DiagnosticApi/TrackingDiagnosticAPI.cs [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+            lock (_responsesToIgnore)
+            {
+                _responsesToIgnore.Add(value);
+                _diagnosticApi.SendIfConnected(_setPayloadFunc(_value, _diagnosticApi.connectedDeviceID));
+            }
+        }
+
+        public void HandleResponse(Result<TData> result)
+        {
+            // TODO: Do something with any errors?
```
Comment (DavidFee-UL) [inline TF_Service_dotNet/TouchFree/Connections/DiagnosticApi/TrackingDiagnosticAPI.cs]: Resolve or create a new ticket in the backlog and reference here

File: TF_Service_dotNet/TouchFree/Connections/DiagnosticApi/TrackingDiagnosticAPI.cs [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+            configManager.TrackingConfig = trackingConfigurationToStore;
+            TrackingConfigFile.SaveConfig(trackingConfigurationToStore);
+        }
+    }
+
+    private async Task MessageQueueReader()
+    {
+        while (true)
+        {
+            await Task.Delay(10); // TODO: Use threading timer instead of loop with delay to reduce allocations
```
Comment (DavidFee-UL) [inline TF_Service_dotNet/TouchFree/Connections/DiagnosticApi/TrackingDiagnosticAPI.cs]: Resolve or create a new ticket in the backlog and reference here

File: TF_Service_dotNet/TouchFree/Connections/DiagnosticApi/TrackingDiagnosticAPI.cs [hunk 3]
```diff
@@ -0,0 +1,562 @@
...
+                                {
+                                    // Set initialized variables to the device, get uninitialized variables
+                                    allowImages.Match(RequestSetAllowImages, RequestGetAllowImages);
+                                    analyticsEnabled.Match(RequestSetAnalyticsMode, RequestGetAnalyticsMode);
+                                    cameraReversed.Match(RequestSetCameraOrientation, RequestGetCameraOrientation);
+                                    maskingData.Match(RequestSetImageMask, RequestGetImageMask);
+                                    RequestGetDeviceInfo();
+                                }
+                            }
+                        }); // TODO: Handle failure?
```
Comment (DavidFee-UL) [inline TF_Service_dotNet/TouchFree/Connections/DiagnosticApi/TrackingDiagnosticAPI.cs]: Resolve these or create a new ticket in the backlog and reference here

File: TF_Service_dotNet/TouchFree/Connections/MessageQueues/TrackingApiChangeQueueHandler.cs [hunk 4]
```diff
@@ -79,199 +82,132 @@ protected override Result<Empty> ValidateContent(IncomingRequestWithId request)
...
-
-                    if (response.needsAnalytics)
-                    {
-                        response.state.analyticsEnabled = boolResponse;
-                    }
-
-                    clientMgr.SendResponse(response.state, ActionCode.TRACKING_STATE);
-                }
+                    requestID = requestId,
+                    // TODO: Errors for cases with no value when caused by an error. Should be refactored to use Result
```
Comment (DavidFee-UL) [inline TF_Service_dotNet/TouchFree/Connections/MessageQueues/TrackingApiChangeQueueHandler.cs]: Resolve or create a new ticket in the backlog and reference here

File: TF_Service_dotNet/TouchFree/Connections/MessageQueues/TrackingApiChangeQueueHandler.cs [hunk 5]
```diff
@@ -79,199 +82,132 @@ protected override Result<Empty> ValidateContent(IncomingRequestWithId request)
...
-                    value => new SuccessWrapper<bool?>(true, "Analytics State", value),
-                    error => new SuccessWrapper<bool?>(false, error.Message, null));
+                await diagnosticApi.RequestSetAll(data);
 
-                response.needsAnalytics = false;
-                trackingApiResponse = response;
+                TrackingConfigFile.SaveConfig(trackingFromFile);
             }
+
+            // TODO: Handle setting config that hasn't been initialised yet
```
Comment (DavidFee-UL) [inline TF_Service_dotNet/TouchFree/Connections/MessageQueues/TrackingApiChangeQueueHandler.cs]: Resolve or create a new ticket in the backlog and reference here

File: TF_Service_dotNet/TouchFree/Connections/MessageQueues/TrackingApiChangeQueueHandler.cs [hunk 6]
```diff
@@ -79,199 +82,132 @@ protected override Result<Empty> ValidateContent(IncomingRequestWithId request)
...
-            bool needsMask = contentObj.TryGetValue("mask", out maskToken);
-            bool needsImages = contentObj.TryGetValue("allowImages", out allowImagesToken);
-            bool needsOrientation = contentObj.TryGetValue("cameraReversed", out cameraReversedToken);
-            bool needsAnalytics = contentObj.TryGetValue("analyticsEnabled", out analyticsEnabledToken);
-
-            trackingApiResponse = new TrackingResponse(request.RequestId, request.OriginalContent, false, needsMask, needsImages, needsOrientation, needsAnalytics);
-            responseOriginTime = DateTimeOffset.Now.ToUnixTimeMilliseconds();
-
-            var trackingFromFile = configManager.TrackingConfig;
+            // TODO: Almost a repeat of HandleValidationError but the error message goes in a different place. Review this!
```
Comment (DavidFee-UL) [inline TF_Service_dotNet/TouchFree/Connections/MessageQueues/TrackingApiChangeQueueHandler.cs:108]: Resolve or create a new ticket in the backlog and reference here

**Review by GabrielHomewoodUL — APPROVED** [review]

**Pull request merged**